build-scripts/build_tests.py

```diff
@@ -0,0 +1,230 @@
+        -> None:
+    file_contents = ssg.jinja.process_file_with_macros(str(file.absolute()), env_yaml)
+    shared_script_path = shared_output_path / file.name
+    _write_path(file_contents, shared_script_path)
+
+
+def _copy_and_process_shared(env_yaml: dict, output_path: pathlib.Path, root_path: pathlib.Path) \
+        -> None:
+    tests_shared_root = root_path / "tests" / "shared"
+    shared_output_path = output_path / "shared"
+    shared_output_path.mkdir(parents=True, exist_ok=True)
+    for file in tests_shared_root.iterdir():  # type: pathlib.Path
+        # We only support one level deep, this avoids recursive functions
+        if file.is_dir():
+            for sub_file in file.iterdir():
+                shared_output_path_sub = shared_output_path / file.name
+                shared_output_path_sub.mkdir(parents=True, exist_ok=True)
+                _process_shared_file(env_yaml, sub_file, shared_output_path_sub)
+        else:
+            _process_shared_file(env_yaml, file, shared_output_path)
+
+
+def _process_local_tests(benchmark_cpes: Set[str], env_yaml: dict, rule_output_path: pathlib.Path,
+                         rule_tests_root: pathlib.Path) -> None:
+    logger = logging.getLogger()
+    for test in rule_tests_root.iterdir():  # type: pathlib.Path
+        if not test.name.endswith(".sh"):
+            logger.debug("Skipping file %s in rule %s is it doesn't end with .sh.",
+                         test.name, rule_output_path.name)
+            continue
+        if not _is_test_file(test.name):
+            file_contents = ssg.jinja.process_file_with_macros(str(test.absolute()),
+                                                               env_yaml)
+            output_file = rule_output_path / test.name
+            rule_output_path.mkdir(parents=True, exist_ok=True)
+            _write_path(file_contents, output_file)
+        file_contents = test.read_text()
+        scenario = tests.ssg_test_suite.rule.Scenario(test.name, file_contents)
+        if scenario.matches_platform(benchmark_cpes):
+            content = ssg.jinja.process_file_with_macros(str(test.absolute()), env_yaml)
+            rule_output_path.mkdir(parents=True, exist_ok=True)
+            _write_path(content, rule_output_path / test.name)
+
+
+def _process_rules(benchmark_cpes: Set[str], env_yaml: Dict, output_path: pathlib.Path,
+                   rules: Iterable[pathlib.Path], templates_root: pathlib.Path,
+                   product_rules: list) -> None:
+    logger = logging.getLogger()
+    for rule_file in rules:  # type: pathlib.Path
+        rendered_rule_obj = ssg.yaml.open_raw(str(rule_file))
+        rule_path = pathlib.Path(rendered_rule_obj["definition_location"])
+        rule_root = rule_path.parent
+        rule_id = rule_root.name
+        if rule_id not in product_rules:
+            logger.debug("Skipping %s since it is not in the product", rule_id)
```

Comment:
You don't have to open the resolved rule yaml file if it is skipped. You can get the rule id from the file name. It will save some time if you don't open resolved rules that aren't going to be processed.